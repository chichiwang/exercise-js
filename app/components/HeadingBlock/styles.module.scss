@use "sass:map";
@use "app/var.module";

$padding-unit: map.get(var.$content, 'padding');

.headingBlock {
  margin: calc(2 * $padding-unit) 0;
  padding: 0 $padding-unit;
  width: 100%;
}
