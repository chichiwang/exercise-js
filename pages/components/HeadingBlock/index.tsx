import React from 'react';
import classNames from 'classnames';

import { HeadingLevel } from 'app/types';
import Heading from 'pages/components/Heading';

import styles from './styles.module.scss';

interface HeadingBlockProps extends React.PropsWithChildren {
  heading: string,
  id?: string,
  level: HeadingLevel,
  center?: boolean,
  className?: string,
}

function HeadingBlock({
  heading,
  id,
  level,
  center = false,
  children,
  className,
}: HeadingBlockProps) {
  const headingProps = {
    center,
    id,
    level,
  };

  return (
    <section className={classNames(className, styles.headingBlock)}>
      <Heading {...headingProps}>
        {heading}
      </Heading>
      <article>
        {children}
      </article>
    </section>
  );
}

export default HeadingBlock;
