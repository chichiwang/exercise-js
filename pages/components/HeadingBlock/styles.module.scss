@use "sass:map";
@use "app/var.module";

$padding-unit: map.get(var.$content, 'padding');

.headingBlock {
  margin: calc(2 * $padding-unit) auto;
  padding: 0 $padding-unit;
  max-width: map.get(var.$content, 'article-max-width');
}
