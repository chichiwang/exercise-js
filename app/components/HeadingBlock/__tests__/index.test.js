import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import HeadingBlock from '../index';

describe('HeadingBlock Component', () => {
  test('matches uncentered snapshot', () => {
    const tree = renderer.create(
      <HeadingBlock heading="Test Heading" level="1">
        <p>Block content</p>
      </HeadingBlock>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('matches centered snapshot', () => {
    const tree = renderer.create(
      <HeadingBlock heading="Test Heading" level="1" center>
        <p>Block content</p>
      </HeadingBlock>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('matches className snapshot', () => {
    const tree = renderer.create(
      <HeadingBlock className="margin-top-large" heading="Test Heading" level="3">
        <p>Block content</p>
      </HeadingBlock>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders child components', () => {
    const childText = 'Paragraph Contents';

    render(
      <HeadingBlock heading="Test Heading" level="1" center>
        <p>{childText}</p>
      </HeadingBlock>,
    );

    expect(screen.getByText(childText)).toBeInTheDocument();
  });

  describe('heading levels', () => {
    test('contains a h1 tag', () => {
      const headingText = 'H1 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="1" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });

    test('contains a h2', () => {
      const headingText = 'H2 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="2" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });

    test('contains a h3', () => {
      const headingText = 'H3 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="3" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 3 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });

    test('contains a h4', () => {
      const headingText = 'H4 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="4" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 4 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });

    test('contains a h5', () => {
      const headingText = 'H5 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="5" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });

    test('contains a h6', () => {
      const headingText = 'H6 tag contents';
      const content = 'Block content';

      render(
        <HeadingBlock heading={headingText} level="6" center>
          <p>{content}</p>
        </HeadingBlock>,
      );

      expect(screen.getByRole('heading', { level: 6 })).toHaveTextContent(headingText);
      expect(screen.getByText(content)).toBeInTheDocument();
    });
  });
});
