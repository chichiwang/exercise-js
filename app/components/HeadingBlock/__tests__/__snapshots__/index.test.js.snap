// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadingBlock Component matches centered snapshot 1`] = `
<section
  className="headingBlock"
>
  <div
    className="container center"
  >
    <div
      className="wrapper"
    >
      <h1
        className="tag"
      >
        Test Heading
      </h1>
    </div>
  </div>
  <article>
    <p>
      Block content
    </p>
  </article>
</section>
`;

exports[`HeadingBlock Component matches className snapshot 1`] = `
<section
  className="margin-top-large headingBlock"
>
  <div
    className="container"
  >
    <div
      className="wrapper"
    >
      <h3
        className="tag"
      >
        Test Heading
      </h3>
    </div>
  </div>
  <article>
    <p>
      Block content
    </p>
  </article>
</section>
`;

exports[`HeadingBlock Component matches uncentered snapshot 1`] = `
<section
  className="headingBlock"
>
  <div
    className="container"
  >
    <div
      className="wrapper"
    >
      <h1
        className="tag"
      >
        Test Heading
      </h1>
    </div>
  </div>
  <article>
    <p>
      Block content
    </p>
  </article>
</section>
`;
